### DISCUSSION TITLE
Failed to validate JWT signature

### DISCUSSION THREAD

**Discussion by ccannell** [nextauthjs/next-auth#3114] [Help]
### Question 💬

I'm trying to setup NextAuth to use [Intuit](https://developer.intuit.com/app/developer/qbo/docs/develop/authentication-and-authorization/openid-connect) as a custom provider.  I'm getting a failed to validate JWT signature error. The key that I receive is:

`
{"jwk":{"kty":"RSA","e":"AQAB","use":"sig","kid":"OPICPRD09182014","alg":"RS256","n":"<omitted>"},"keyObject":{},"algorithms":{}}
`

and the error is:

> [next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error failed to validate JWT signature {
  error: {
    message: 'failed to validate JWT signature',
    stack: 'RPError: failed to validate JWT signature\n' +
      '    at Client.validateJWT (/Users/ccannell/code/magothy/next-auth-example/node_modules/openid-client/lib/client.js:1010:11)\n' +
      '    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n' +
      '    at async Client.validateIdToken (/Users/ccannell/code/magothy/next-auth-example/node_modules/openid-client/lib/client.js:706:49)\n' +
      '    at async Client.callback (/Users/ccannell/code/magothy/next-auth-example/node_modules/openid-client/lib/client.js:462:7)\n' +
      '    at async oAuthCallback (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/core/lib/oauth/callback.js:103:16)\n' +
      '    at async Object.callback (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/core/routes/callback.js:50:11)\n' +
      '    at async NextAuthHandler (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/core/index.js:103:28)\n' +
      '    at async NextAuthNextHandler (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/next/index.js:40:7)\n' +
      '    at async /Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/next/index.js:80:32\n' +
      '    at async Object.apiResolver (/Users/ccannell/code/magothy/next-auth-example/node_modules/next/dist/server/api-utils.js:102:9)',
    name: 'RPError'
  },
  providerId: 'intuit',
  message: 'failed to validate JWT signature'
}
[next-auth][error][CALLBACK_OAUTH_ERROR] 
https://next-auth.js.org/errors#callback_oauth_error failed to validate JWT signature RPError: failed to validate JWT signature
    at Client.validateJWT (/Users/ccannell/code/magothy/next-auth-example/node_modules/openid-client/lib/client.js:1010:11)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Client.validateIdToken (/Users/ccannell/code/magothy/next-auth-example/node_modules/openid-client/lib/client.js:706:49)
    at async Client.callback (/Users/ccannell/code/magothy/next-auth-example/node_modules/openid-client/lib/client.js:462:7)
    at async oAuthCallback (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/core/lib/oauth/callback.js:103:16)
    at async Object.callback (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/core/routes/callback.js:50:11)
    at async NextAuthHandler (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/core/index.js:103:28)
    at async NextAuthNextHandler (/Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/next/index.js:40:7)
    at async /Users/ccannell/code/magothy/next-auth-example/node_modules/next-auth/next/index.js:80:32
    at async Object.apiResolver (/Users/ccannell/code/magothy/next-auth-example/node_modules/next/dist/server/api-utils.js:102:9) {
  name: 'OAuthCallbackError'

I'm using the next-auth-eample with the following for `[...nextauth].js`

```javascript
import NextAuth from "next-auth"
import * as Fauna from "faunadb"
import { FaunaAdapter } from "@next-auth/fauna-adapter"

const client = new Fauna.Client({
  secret: process.env.FAUNA_SECRET,
  scheme: process.env.FAUNA_SCHEME,
  domain: process.env.FAUNA_DOMAIN,
  port: 8443,
})

// For more information on each option (and a full list of options) go to
// https://next-auth.js.org/configuration/options
export default NextAuth({
  // https://next-auth.js.org/configuration/providers
  providers: [
    {
      id: "intuit",
      name: "Intuit",
      clientId: process.env.INTUIT_ID,
      clientSecret: process.env.INTUIT_SECRET,
      type: "oauth",
      version: "2.0",
      wellKnown: process.env.INTUIT_DOMAIN,
      authorization: { params: { scope: "openid email profile com.intuit.quickbooks.accounting" } },
      idToken: true,
      checks: ["pkce", "state"],
      profile(profile) {
        return {
          id: profile.sub,
          name: profile.name,
          email: profile.email,
          image: profile.picture,
        }
      },
    },
    
  ],

  adapter: FaunaAdapter({ faunaClient: client}),

  // The secret should be set to a reasonably long random string.
  // It is used to sign cookies and to sign and encrypt JSON Web Tokens, unless
  // a separate secret is defined explicitly for encrypting the JWT.
  secret: process.env.SECRET,

  session: {
    // Use JSON Web Tokens for session instead of database sessions.
    // This option can be used with or without a database for users/accounts.
    // Note: `jwt` is automatically set to `true` if no database is specified.
    jwt: true,

    // Seconds - How long until an idle session expires and is no longer valid.
    // maxAge: 30 * 24 * 60 * 60, // 30 days

    // Seconds - Throttle how frequently to write to database to extend a session.
    // Use it to limit write operations. Set to 0 to always update the database.
    // Note: This option is ignored if using JSON Web Tokens
    // updateAge: 24 * 60 * 60, // 24 hours
  },

  // JSON Web tokens are only used for sessions if the `jwt: true` session
  // option is set - or by default if no database is specified.
  // https://next-auth.js.org/configuration/options#jwt
  jwt: {
    // A secret to use for key generation (you should set this explicitly)
    // secret: 'INp8IvdIyeMcoGAgFGoA61DdBglwwSqnXJZkgz8PSnw',
    secret: process.env.SECRET,
    // Set to true to use encryption (default: false)
    encryption: false,
    // You can define your own encode/decode functions for signing and encryption
    // if you want to override the default behaviour.
    // encode: async ({ secret, token, maxAge }) => {},
    // decode: async ({ secret, token, maxAge }) => {},
  },

  // You can define custom pages to override the built-in ones. These will be regular Next.js pages
  // so ensure that they are placed outside of the '/api' folder, e.g. signIn: '/auth/mycustom-signin'
  // The routes shown here are the default URLs that will be used when a custom
  // pages is not specified for that route.
  // https://next-auth.js.org/configuration/pages
  pages: {
    // signIn: '/auth/signin',  // Displays signin buttons
    // signOut: '/auth/signout', // Displays form with sign out button
    // error: '/auth/error', // Error code passed in query string as ?error=
    // verifyRequest: '/auth/verify-request', // Used for check email page
    // newUser: null // If set, new users will be directed here on first sign in
  },

  // Callbacks are asynchronous functions you can use to control what happens
  // when an action is performed.
  // https://next-auth.js.org/configuration/callbacks
  callbacks: {
    // async signIn({ user, account, profile, email, credentials }) { return true },
    // async redirect({ url, baseUrl }) { return baseUrl },
    // async session({ session, token, user }) { return session },
    // async jwt({ token, user, account, profile, isNewUser }) { return token }
  },

  // Events are useful for logging
  // https://next-auth.js.org/configuration/events
  events: {},

  // You can set the theme to 'light', 'dark' or use 'auto' to default to the
  // whatever prefers-color-scheme is set to in the browser. Default is 'auto'
  theme: {
    colorScheme: "light",
  },

  // Enable debug messages in the console if you are having problems
  debug: true,
})
```

I assume I've configured NextAuth incorrectly but I'm not sure how. Thank you.

**Comment by GuiSim** [comment]
@ccannell did you end up figuring it out? I'm encountering the same issue.

EDIT: it looks like the JWK used by Intuit has 1024 bits and Jose requires 2048 min for RSA
https://oauth.platform.intuit.com/op/v1/jwks

This is the check that fails
https://github.com/panva/jose/blob/404818b19c4e62a97b170ab319dbec1135402890/src/runtime/node/check_modulus_length.ts#L70

Setting `idToken: false` doesn't solve it
`'id_token detected in the response, you must use client.callback() instead of client.oauthCallback()'`

**Reply by GuiSim** [reply]
Follow up discussion with openid-client https://github.com/panva/node-openid-client/discussions/492

**Comment by fasilminale** [comment]
@ccannell @GuiSim did you solve this issue?

**Comment by fasilminale** [comment]
{
      clientId: process.env.QUICK_BOOKS_CLIENT_ID,
      clientSecret: process.env.QUICK_BOOKS_CLIENT_SECRET,
      id: 'quickbooks',
      name: 'QuickBooks',
      type: 'oauth',
      authorization: {
        url: 'https://appcenter.intuit.com/connect/oauth2',
        params: { scope: 'com.intuit.quickbooks.accounting openid profile email phone address' }
      },
      token: {
        url: 'https://oauth.platform.intuit.com/oauth2/v1/tokens/bearer'
      }
      //   idToken: true,
      //   checks: ['pkce', 'state'],
      //   profile(profile) {
      //     return {
      //       id: profile.sub,
      //       name: profile.name,
      //       email: profile.email,
      //       image: profile.picture
      //     }
      //   }
    }

**Reply by GuiSim** [reply]
Solved in a not great way here: https://github.com/nextauthjs/next-auth/issues/4554#issuecomment-1175486720

**Reply by fasilminale** [reply]
Thanks @GuiSim for replying. I patched it but i'm still getting this error

```javascript
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error failed to validate JWT signature {
  error: RPError: failed to validate JWT signature
      at Client.validateJWT (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/openid-client/lib/client.js:1055:11)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      at async Client.validateIdToken (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/openid-client/lib/client.js:745:49)
      at async Client.callback (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/openid-client/lib/client.js:488:7)
      at async oAuthCallback (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/next-auth/core/lib/oauth/callback.js:127:16)
      at async Object.callback (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/next-auth/core/routes/callback.js:52:11)
      at async NextAuthHandler (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/next-auth/core/index.js:201:28)
      at async NextAuthNextHandler (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/next-auth/next/index.js:23:19)
      at async /Users/addisaba/Documents/projects/iwork/rpa/node_modules/next-auth/next/index.js:59:32
      at async Object.apiResolver (/Users/addisaba/Documents/projects/iwork/rpa/node_modules/next/dist/server/api-utils.js:102:9) {
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'quickbooks',
  message: 'failed to validate JWT signature'
}
```

this is the provider object
```javascript
{
      clientId: process.env.QUICK_BOOKS_CLIENT_ID,
      clientSecret: process.env.QUICK_BOOKS_CLIENT_SECRET,
      id: 'quickbooks',
      name: 'QuickBooks',
      type: 'oauth',
      wellKnown: 'https://developer.api.intuit.com/.well-known/openid_sandbox_configuration',
      authorization: { params: { scope: 'com.intuit.quickbooks.accounting openid profile email phone address' } },
      idToken: true,
      checks: ['pkce', 'state']
      //   profile(profile) {
      //     return {
      //       id: profile.sub,
      //       name: profile.name,
      //       email: profile.email,
      //       image: profile.picture
      //     }
      //   }
    }
```

**Reply by fasilminale** [reply]
I edited this file.

<img width="1638" alt="Screen Shot 2022-12-01 at 12 28 17 PM" src="https://user-images.githubusercontent.com/30689833/205016561-fbccaa97-8fbd-47a8-ab92-bf33120bff7a.png">

**Reply by fasilminale** [reply]
I can't believe it finally works. Make sure to restart the server. Thank you @GuiSim

**Reply by sammykumar** [reply]
fyi, i had to delete the .next folder to get my patched jose dependency to take effect. I'm sure stuff has changed with how nextjs compiles so for whatever reason, simply restarting the server doesn't take your patched files into effect anymore.
